src/python_testing/modebase_cluster_check.py

```diff
@@ -0,0 +1,159 @@
+        - The Mode values of all entries are unique.
+        - The Label values of all entries are unique.
+
+        Args:
+          endpoint: The endpoint used for the requests to the cluster.
+
+        Returns:
+          A list of ModeOptionStruct supported by the cluster.
+        """
+        # Get the supported modes
+        supported_modes = await self.read_single_attribute_check_success(endpoint=endpoint,
+                                                                         cluster=self.cluster,
+                                                                         attribute=self.attributes.SupportedModes)
+
+        # Check if the list of supported modes is larger than 2
+        asserts.assert_greater_equal(len(supported_modes), 2, "SupportedModes must have at least 2 entries!")
+        # Check that supported modes are less than 255
+        asserts.assert_less_equal(len(supported_modes), 255, "SupportedModes must have at most 255 entries!")
+
+        # Check for repeated labels or modes
+        labels = set()
+        modes = set()
+        for mode_option_struct in supported_modes:
+            # Verify that the modes in all ModeOptionStruct in SupportedModes are unique.
+            if mode_option_struct.mode in modes:
+                asserts.fail("SupportedModes can't have repeated Mode values")
+            else:
+                modes.add(mode_option_struct.mode)
+            # Verify that the labels in all ModeOptionStruct in SupportedModes are unique.
+            if mode_option_struct.label in labels:
+                asserts.fail("SupportedModes can't have repeated Label values")
+            else:
+                labels.add(mode_option_struct.label)
+
+        return supported_modes
+
+    def check_tags_in_lists(self, supported_modes, required_tags=None):
+        """ Validates the ModeTags values.
+
+        This function evaluates the ModeTags of each ModeOptionStruct:
+        - Should have at least one tag.
+        - Should be maximum 16bits in size.
+        - Should be a Mfg tag or one of the supported ones (either common or specific).
+        - Should have at least one common or specific tag.
+        - If defined, verify that at least one of the "required_tags" exists.
+
+        Args:
+          supported_modes: A list of ModeOptionStruct.
+          required_tags: List of tags that are required according to the cluster spec.
+        """
+        # Verify the ModeTags on each ModeOptionStruct
+        for mode_option_struct in supported_modes:
+            # Shuld have at least one entry
+            if len(mode_option_struct.modeTags) == 0:
+                asserts.fail("The ModeTags field should have at least one entry.")
+
+            # Check each ModelTag
+            at_least_one_common_or_derived = False
+            for tag in mode_option_struct.modeTags:
+                # Value should not larger than 16bits
+                if tag.value > MAX_MODE_TAG or tag.value < 0:
```

Comment:
nit: you use the range below for is_mfg, so maybe this shows the range clearer:

```suggestion
                if not (0 <= tag.value <= MAX_MODE_TAG):
```